### ISSUE TITLE
[x86] codegen for fcmp oeq is inconsistent

### ISSUE THREAD

**Issue by rotateright** [llvm/llvm-project#33911]
|  |  |
| --- | --- |
| Bugzilla Link | [34563](https://llvm.org/bz34563) |
| Version | trunk |
| OS | All |
| CC | @topperc,@RKSimon |

## Extended Description
```
bool fcmp_oeq(double f1, double f2) {
  return f1 == f2;
}

bool fcmp_oeq_twice(double f1, double f2, double f3, double f4) {
  return f1 == f2 && f3 == f4;
}
```
Or as IR:
```
define i1 @fcmp_oeq(double %f1, double %f2) {
  %cmp = fcmp oeq double %f1, %f2
  ret i1 %cmp
}

define i1 @fcmp_oeq_twice(double %f1, double %f2, double %f3, double %f4) {
  %cmp1 = fcmp oeq double %f1, %f2
  %cmp2 = fcmp oeq double %f3, %f4
  %and = and i1 %cmp1, %cmp2
  ret i1 %and
}
```
----------------------------------------------------------------------------
```
$ ./llc -o - -mtriple=x86_64-unknown-unknown fcmps.ll

fcmp_oeq(double, double):                          # @fcmp_oeq(double, double)
        cmpeqsd %xmm1, %xmm0
        movq    %xmm0, %rax
        andl    $1, %eax
        retq
fcmp_oeq_twice(double, double, double, double):                 # @fcmp_oeq_twice(double, double, double, double)
        ucomisd %xmm1, %xmm0
        setnp   %al
        sete    %cl
        andb    %al, %cl
        ucomisd %xmm3, %xmm2
        setnp   %dl
        sete    %al
        andb    %dl, %al
        andb    %cl, %al
        retq
```
-----------------------------------------------------------------------------

x86 doesn't have a 'setcc' for oeq (?!), so if we're using 'ucomisd', we have to do an and-of-setcc to generate that predicate. If we use 'cmpeqsd' as in the first example, we incur a vector-to-scalar register move. That might not be as fast?

The inconsistency here should be investigated. But it's also possible that we're doing the wrong thing for both cases. In the 2nd example if we use 'cmpeqsd', then we could reduce the instruction count with something like:
```
        cmpeqsd %xmm1, %xmm0
        cmpeqsd %xmm3, %xmm2
        andps   %xmm0, %xmm2
        movd    %xmm0, %eax
        andl    $1, %eax
```

**Comment by RKSimon** [comment]
pre-AVX is missing support in convertIntLogicToFPLogic due to this:
```
  // The vector ISA for FP predicates is incomplete before AVX, so converting
  // COMIS* to CMPS* may not be a win before AVX.
  // TODO: Check types/predicates to see if they are available with SSE/SSE2.
  if (!Subtarget.hasAVX() || VT != MVT::i1 || N0.getOpcode() != ISD::SETCC ||
      !N0.hasOneUse() || !N1.hasOneUse())
    return SDValue();
```

**Comment by RKSimon** [comment]
Current Codegen: https://reviews.llvm.org/D121210

**Comment by RKSimon** [comment]
AVX code was fixed by 09e71c367af36a07cce04a812dcd9c62bebfbcdf
SSE code was fixed by e4ab2024a65a2233d258fc3377bd6c84376376aa
